### PR TITLE
refactor: execute compaction plan

### PR DESCRIPTION
**Pull request by NGA-TRAN**
A part of  https://github.com/influxdata/idpe/issues/16922

When this PR is merge, compactor2 will compact all L0 and L1 files of a partition in one run. We expect if there are a lot of files in a partition, this partition will get error and added into `skipped_compaction`. We expect some unexpected things happen with this first run, too.

We have NOT looked into `skipped_compactions` to ignore partitions in there. There is an todo for its next

### CODE REVIEW

**Review by pauldix — APPROVED** [review]
I haven't been close enough to everything to know all the details, but for the most part this makes sense to me. Hitting approve for now so we can see what this does over the weekend after it automatically gets deployed. Will let @crepererum provide the real review on Monday.

File: compactor2/src/driver.rs [hunk 1]
```diff
@@ -12,9 +12,24 @@ use crate::{
...
+//    . Level-N+1 files are results of compacting Level-N and/or Level-N+1 files, their `created_at`
+//      can be after the `created_at` of other Level-N files but they may include data loaded before
+//      the other Level-N files. Hence we should never use `created_at` of Level-N+1 files to order
+//      them with Level-N files.
+//    . We can only compact different sets of files of the same partition concurrently into the same target_level.
 pub async fn compact(config: &Config, components: &Arc<Components>) {
     let partition_ids = components.partitions_source.fetch().await;
 
+    // TODO: https://github.com/influxdata/influxdb_iox/issues/6657
+    // either here or before invoking this function to ignore this partition if it is in skipped_compactions
```
Comment (NGA-TRAN) [inline compactor2/src/driver.rs:31]: @crepererum : this is important to do next if you are online before me.

File: compactor2/src/components/compact/compact_builder.rs [hunk 2]
```diff
@@ -24,13 +24,19 @@ use super::partition::PartitionInfo;
...
     CompactLogicalPlan {
         source: iox_query::frontend::reorg::Error,
     },
+
+    #[snafu(display("Error building compact physical plan  {}", source))]
+    CompactPhysicalPlan { source: DataFusionError },
 }
 
+/// Observer function (for testing) that is invoked on the physical plan to be run
+type PlanObserver = Box<dyn Fn(&dyn ExecutionPlan) + Send>;
```
Comment (crepererum) [inline compactor2/src/components/compact/compact_builder.rs:38]: This is exactly what components are for. The actual execution should be another component that can also implement wrapping / observation.

File: compactor2/src/driver.rs [hunk 3]
```diff
@@ -106,3 +152,125 @@ pub async fn compact(config: &Config, components: &Arc<Components>) {
         .collect::<()>()
         .await;
 }
+
+#[cfg(test)]
```
Comment (crepererum) [inline compactor2/src/driver.rs:156]: These are basically "component integration tests" and should probably life in a dedicated module.

**Pull request merged**